clang/lib/CodeGen/CGExprScalar.cpp

```diff
@@ -2375,10 +2375,12 @@ Value *ScalarExprEmitter::VisitCastExpr(CastExpr *CE) {
     llvm::Type *DstTy = ConvertType(DestTy);
 
     if (SrcTy->isPointerTy() && DstTy->isPointerTy() &&
-        SrcTy->getPointerAddressSpace() != DstTy->getPointerAddressSpace())
+        SrcTy->getPointerAddressSpace() != DstTy->getPointerAddressSpace()) {
       Src = Builder.CreateAddrSpaceCast(
           Src,
           llvm::PointerType::get(VMContext, DstTy->getPointerAddressSpace()));
+      SrcTy = Src->getType();
```

Comment:
Could you please clarify why this change is needed and why it was not needed before?

Reply:
The community change changes sret address space to that of the alloca address space, instead of the address space of the pointee.  For the test case in CodeGenCXX/sret_cast_with_nonzero_alloca_as.cpp, this results in an AS of 0 instead of 4.   (We cast it away in this case, but we didn't need to before.)

Reply:
> The thing is, we adjusted the Src expression above, but the SrcTy still refers to the unadjusted Src expression.


Poping years after the patch, but this adjustment doesn't look right here. It should probably be handled in the AST by adding the CK_AddressSpaceConversion or when we codegen the builtin by returning the cast address space. I think the 2nd option is more right as it is applying the SYCL address space deduction rules (the first makes assumption on the target address spaces in the AST).

Reply:
We may leave this "as-is" to resolve failures in pulldown and capture this

> It should probably be handled in the AST by adding the CK_AddressSpaceConversion or when we codegen the builtin by returning the cast address space. I think the 2nd option is more right as it is applying the SYCL address space deduction rules (the first makes assumption on the target address spaces in the AST).

in a separate tracker.

The overall change of sret address space doesn't seem like a concern. It is this snippet of code is... strange.

Reply:
> We may leave this "as-is" to resolve failures

Oh absolutely, my comment is clearly out of scope of the PR ;)

Reply:
Were you referring to the community FIXME below as "strange" or something else?

I can file a tracker for a proper fix later on, so we can make progress here.   If you and @Naghasan can add specifics to it, I would appreciate it.

Reply:
> Were you referring to the community FIXME below as "strange" or something else?

I was referring to the downstream code.

Reply:
No, the `CreateAddrSpaceCast` above (added in '22). I agree with the comment below, an address space cast shouldn't happen here and I think this one is avoidable.